### PR TITLE
0.16.0

### LINKED ISSUE #422: Don't lock render before calling view function.
Recently I have been playing around with modules that map external state of things like `mouse`, `scroll` and `location` to hyperapp state. I have observed that it certain cases when `update` is called in rapid succession; that the data used in the last render is not the data in the state.

If I remove the skip render check (but leave rAF), the behaviour goes away:
```
function repaint() {
  if (props.view) { requestAnimationFrame(render) }
}
```

I started looking at [`requestAnimationFrame`](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame) as I haven't used it since I experimented with WebGL years ago. From what I can gather:

- Requests that the browser call a specified function before the next repaint
- You should call this method whenever you're ready to update your view onscreen

It doesn't say anything about skipping renders 🤔 and it makes me wonder why we are doing this? I have always thought of `requestAnimationFrame` as like a smart debounce/throttle which you throw whatever you want at, and it makes sure not to over compute.

Looking at the current setup it seems that a state update `assign(state, data)` could be made, which calls `repaint()` when `skipRender` is true meaning that `render` is never called. This would explain the behaviours I have witnessed.

Interestingly removing all trace of `skipRender` seems to have no adverse effects at all on hyperapp even when challenged with rapid state updates. Do we need it for some reason I am not remembering?

### LINKED ISSUE #425: The state should be immutable
# Current state
Since the addition of modules, the state is no longer immutable: when an action, or a nested action, or an action in a module is called, its result is merged with the current state (or state slice for nested actions/modules).

This prevent users from using memoize functions on the view with simple `===` checks for equality checks and forces to use deep equality.

For example, the following pen should work with immutable state: https://codepen.io/mytrill/pen/WZmVWQ?editors=0010

I believe for large apps, the benefit of having very cheap memoize functions based on immutable state far outweight the cost of maintaining an immutable state (both in terms of performance and source code size).

# Expected behavior
When an action gets executed, the result that it returns should get merged to the state slice/module's state this action operates on and create a new instance of the state and re-use the part of the state that hasn't been touched.
Here is an example: https://upload.wikimedia.org/wikipedia/commons/thumb/5/56/Purely_functional_tree_after.svg/438px-Purely_functional_tree_after.svg.png

In this case, when the node "f" is modified ("e" gets added to it) the update function recursively goes up the state tree, creating new nodes when modified and re-using un-touched nodes.

## Advantages
 - Allow debug tools to keep the state history and implement time travel
 - Allow very performant memoization of components/views on large app (deep equality on state can be checked with `===`)

## Drawbacks
 - small performance penalty when executing nested action (need to duplicate a whole branch of the state-tree
 - extra logic/bundle size

# Implementation
PR #426 addresses this issue

### LINKED ISSUE #432: Remove update & thunks.
Obviously this issue needs more info, but the core idea is to remove the concept of update and thunks unless we can demonstrate that using only actions is not enough to do everything that can be accomplished with actions _and_ thunks.

I'll update this issue with a bit of history about why thunks were introduced and why I think they are dispensable now that events and mixins are not part of our architecture.

An counter-alternative to this proposal: https://github.com/hyperapp/hyperapp/issues/433

EDIT: This is **just** a proposal and it may not happen! A different proposal may come out of this or nothing at all. We are just brooding together! ☝️🤓❤️

### PR DESCRIPTION
**Pull request by JorgeBucaran**
- Remove init in favor of using the app() returned actions. 
- Immutable state. Close #425. Thanks @Mytrill. 🎉 
- No-thunks. Close #432.
- Curry actions' data argument. Thanks @Swizz. 🎉 
- Allow calling actions in the view. Close #422.
- Actions now return what the action implementation returns.

### CODE REVIEW

**Comment by okwolf** [comment]
Size impact?

**Comment by zaceno** [comment]
@okwolf @JorgeBucaran  Yes all PR's are gonna need a fix to #442. @SahAssar has the fix in his: https://github.com/hyperapp/hyperapp/pull/441 (that's why his PR has a green check, and ours don't 😄

**Comment by JorgeBucaran** [comment]
@okwolf 1499 B

**Comment by zaceno** [comment]
@JorgeBucaran It grew?! I thought removing thunks would make it much smaller? Oh but immutable state... Is that where the bulk of the increase is?

**Comment by JorgeBucaran** [comment]
@zaceno Yup.

File: src/app.js [hunk 1]
```diff
@@ -49,54 +41,89 @@ export function app(props, container) {
...
+                actions
+              )) === "function"
+            ) {
+              state = state(data)
+            }
+
+            if (
+              state &&
+              !state.then &&
+              state !== (data = getObject(path, globalState))
```
Comment (Mytrill) [inline src/app.js]: No more getters? :(

Comment (JorgeBucaran) [reply src/app.js]: 🤕 No more getters.

File: src/app.js [hunk 2]
```diff
@@ -151,11 +178,11 @@ export function app(props, container) {
       var oldValue = i === "value" || i === "checked" ? element[i] : oldProps[i]
 
       if (value !== oldValue) {
-        setElementProp(element, i, value, oldValue)
+        value !== oldValue && setElementProp(element, i, value, oldValue)
```
Comment (lukejacksonn) [inline src/app.js]: Double checking for `value !== oldValue` this was perviously fixed by https://github.com/hyperapp/hyperapp/pull/429

Comment (JorgeBucaran) [reply src/app.js]: 😅 Whoops. Thanks!

File: src/app.js [hunk 3]
```diff
@@ -1,37 +1,29 @@
...
-  repaint(flush(init(props, (globalState = {}), (globalActions = {}))))
+  repaint(init(props, globalState, globalActions, []))
 
   return globalActions
 
   function repaint() {
-    if (props.view && !skipRender) {
-      requestAnimationFrame(render, (skipRender = !skipRender))
+    if (props.view && !lock) {
+      setTimeout(render, (lock = !lock))
```
Comment (SkaterDad) [inline src/app.js]: What was the reasoning for changing `requestAnimationFrame` to `setTimeout`?

File: src/app.js (hunk 1 above)
Comment (Swizz) [reply src/app.js]: You can make a getter that return a state slice without triggering a render isn't it ?

File: src/app.js (hunk 3 above)
Comment (Swizz) [reply src/app.js]: From a discussion on slack @lukejacksonn found some drawbacks to `requestAnimationFrame` (not working on hidden tabs) plus the fact we needed to mock it for testing.

File: src/app.js (hunk 1 above)
Comment (JorgeBucaran) [reply src/app.js]: @Swizz Yes.

File: src/app.js (hunk 3 above)
Comment (JorgeBucaran) [reply src/app.js]: setTimeout does not have that problems and still allow us to "debounce" re-renders after calling multiple actions in rapid-fire.

File: src/app.js [hunk 4]
```diff
@@ -49,54 +41,89 @@ export function app(props, container) {
...
-            ? data(update)
-            : update(data)
+    function create(state, actions, from, path) {
+      Object.keys(from || {}).map(function(i) {
+        if (typeof from[i] === "function") {
+          actions[i] = function(data) {
+            if (
+              typeof (state = from[i](
+                getObject(path, globalState),
+                actions
```
Comment (Swizz) [inline src/app.js]: Is this the slice actions ?

Comment (JorgeBucaran) [reply src/app.js]: Yes.

**Pull request merged**